learn/appendix_opengl/KHR_debug.md

```diff
@@ -0,0 +1,194 @@
+KHR_debug Extension
+===================
+`KHR_debug` is an extension that provides additional functions to the OpenGL API
+which aids development and debugging of OpenGL applications. This extension is
+available on many drivers and has been core since OpenGL version 4.3. Please
+read the previous article on the topic, [Debug Callback](debug_callback.md),
+before reading this as it provides better context into this extension.
+
+For versions of OpenGL later than 4.3 or if the extension is present you can
+directly use the function as defined in the `GL` class. For OpenGLES contexts
+use the `GL.Khr` class. The subclass version is a drop-in replacement for the
+core version for ES contexts. This extension only works when the context has
+the "Debug" flag set.
+
+> [!NOTE]
+> The `length` parameter for most of the functions given are intended for C
+> users and is in ASCII characters, not UTF-16 characters. Prefer to use C#
+> string API and pass `-1` for the length instead.
```

Comment:
I would say, don't pass `-1`, pass the actual length of the string. `-1` indicates a null-terminated string, and while that is fine as we marshal the string, I find it completely unecessary to force the driver to do a `strlen`.

Reply:
Also this would probably be more appropriate after the first mention of a function that has a `length` parameter as readers might not know what that is yet.

Reply:
Passing a proper length gets complicated as you delve into non-latin code pages, as C# strings are unfortunately UTF-16. This leads to an issue where the length of the string in C# is not the length of the `char*`.

The most correct way to do this would be to use `System.Text.Encoding.UTF8.GetByteCount()`, which would still be a `strlen` except more complicated. I don't think there is a neater way to showcase this correctly without scaring off newbies with C-C# interoperability fluff, other than telling people to use `-1`.

Reply:
Hmm, I guess `-1` is fine. But I do think we should say something about encoding stuff in that case.